### ISSUE TITLE
Reproducing the Amago Performance on PopGym MineSweeperMedium

### ISSUE THREAD

**Issue by Leiay** [UT-Austin-RPL/amago#79]
HI, thank you for your great work and open-sourced code.

We are trying to reproduce your performance on the MineSweeperMedium, on the PopGym dataset. First, we just run the `03_popgym_suite.py` code, without changing any hyper-parameters, but it results in the pink run in the attached figure, which barely exceeds -0.32.

Later, we also try to change the hyper-parameters according to the hyparam table (Table 3 and Table 4) in the amago paper. We made the following changes:
- Changing the timestep encoder from input dim-256-256-256 to input dim-512-512-200
- Gradient norm clipping from 2 to 1
- Learning rate from 3e-4 to 1e-4
- Buffer size from 20k to 80k.
- Exploration step annealing from 400k to 1m.

However, this results in the teal run in the attached figure. We wonder if you could provide an example code to run to reproduce your result in the amago paper, that reaches 0.06 at 18M timesteps? For instance, what are the `max_seq_len`, `train_timesteps_per_epoch`? Are they the default values in the repo?

Thanks very much for your help in advance!

![Image](https://github.com/user-attachments/assets/3d969172-541b-49c6-b04b-46606fd6b7bb)

**Comment by jakegrigsby** [comment]
Ok sorry for the delay.


**TL;DR: There are some bad default hparams now and fixing them gets most of the performance back. I will update the code after a few more experiments. I can't match the old version perfectly yet, but this decline doesn't show up in any other POPGym envs i've tried since the original paper came out.**

<br> <br>

Long version:

1. As a reference, curves from the original seeds.

<img width="300" alt="Image" src="https://github.com/user-attachments/assets/7caee3db-b493-4988-9b81-afdbb7ff6359" />

<img width="300" alt="Image" src="https://github.com/user-attachments/assets/46184239-9c8d-4f92-8381-bcd9fc5e512a" />


In the spreadsheet used to make the original POPGym main text results figure I have

<img width="250" alt="Image" src="https://github.com/user-attachments/assets/8630c42f-62c0-4ef9-a073-1531beabd763" />

at 15M steps

2. Performance on the current script has dropped. Here is a list of things that changed, and my conclusion on whether they matter based on trying them in various combinations. Small sample sizes, but worth noting that the original popgym results were minor changes to the defaults I guessed on a few tries on ConcentrationEasy and went with because we were so ahead of the baselines on the recall-heavy envs. POPGym had unusually low seed variance (with a few exceptions, but MineSweeper isn't one of them)

- The paper incorrectly reports the critic `tau` as .003, when wandb says it was .0025.

- Tstepencoder hidden dim was 512 --> now 256, output dim was 200 --> now 256, Learning rate 1e-4 --> 3e-4, grad clip 1 --> 2. All of these are changes from a recent PR #67 intended to speed up training, but I didn't check MineSweeper when I made them. Easy to roll back.

- popgym used to have a timer in the observation. It was `time / horizon`, where horizon was set to some large number so the script worked on all of the POPGym envs. #67 brought the timer back but forgets to divide by a horizon. Shouldn't matter because the policy normalizes its inputs, but this has now been fixed just to be safe.

- The CRR binary filter in the original results may have been `bool(advantage > 1e-3)` instead of `bool(advantage > 0)`. Definitely shouldn't matter, but I tried it to be safe, and the filter metrics on wandb agree. 

- During a large update I removed learnable (`nn.Embedding`) pos embs in the Transformer and defaulted to the older sin/cos. Experiments said this didn't change performance on the easier POPGym envs, but I've now brought back learnable pos embs as an option.

- The initialization of the SigmaReparam transformer layers has been changed to improve stability. I tried the old way.

- The learning rate warmup used to set in terms of epochs. when I changed this the default warmup became fewer grad steps. I tried increasing it.

- Exploration schedule used to use a long horizon to approximate standard epsilon-greedy. Now there is a standard epsilon greedy option - no approximation needed. I checked three different anneal rates.


<img width="275" alt="Image" src="https://github.com/user-attachments/assets/36a4a0c8-b902-4d68-8ea9-2376a317f3e1" />

From this chaos, the general pattern was that once I fixed the easy hparam changes, we get back above -.15. If there's a difference from the other changes, it's too small to measure with 1 seed, and probably not enough to get back to .02 anyway.

<img width="300" alt="Image" src="https://github.com/user-attachments/assets/e97e7809-9c76-4b50-81aa-166e80959b49" />

3. MineSweeperHard has declined too, even with the change in hparams fixed.


<img width="200" alt="Image" src="https://github.com/user-attachments/assets/029f2119-f233-4f17-9a39-97159565c3d4" />


4. As I said, I test on popgym a lot for debugging. I've never seen a significant decline in results aside from this issue, and checked a few more games with these changes just to safe:

<img width="300" alt="Image" src="https://github.com/user-attachments/assets/48c3eaed-3e51-4096-aff6-34b6431b368d" />

So the damage is hopefully limited to MineSweeper. I'll play around with a few more attempts over the next few days and then decide what to do about this. I will at least update the script to correct the broken hparams, add learnable pos embs, and possibly re-run a sweep of all the envs to get an updated reference score if I can't fix it.

**Comment by Leiay** [comment]
Thank you so much for your detailed exploration and thoughtful investigation!

So far, we’ve tried autoencoder easy, battleship hard, and now minesweeper medium—this is the third environment we’ve explored. We truly appreciate your time and effort in looking into this, as we are also very curious about what makes minesweeper medium perform so well.

From what I understand based on your findings, applying the following hyperparameter adjustments:
- critic tau = 0.0025
- transformer hidden dim = 512, output dim = 200
- learning rate = 1e-4
- grad clip = 1

seems to restore performance to around -0.15, which is very helpful to know. I’ll make sure to reproduce this on my end as well.

That said, there still appears to be a gap in fully recovering the previously reported 0.02, possibly due to changes made afterward, making it difficult to trace back entirely. But in any case, it is reassuring to confirm that this discrepancy is not due to a re-implementation issue on our side.

We really appreciate your patience and dedication in working through this, and we look forward to seeing whether this gap can be fully resolved. If not, it seems reasonable that this behavior might be specific to minesweeper medium.

Thanks again for all your help and for taking the time to share your findings with us! We truly appreciate it and look forward to any further updates you may have.

**Comment by Leiay** [comment]
Hi, 

I tried following the hyperparameter adjustments I mentioned above and ran the command:
```shell
python 03_popgym_suite.py --env MineSweeperMedium --parallel_actors 24 --epochs 650 --dset_max_size 80_000 --env_mode sync --memory_layers 3 --memory_size 256 --run_name <your_run_name> --buffer_dir <your_buffer_dir>
```

However, I’m only reaching **-0.26 at 12M** timesteps, which seems lower than expected.

I checked the differences between my run and your original runs and noticed that the gradient norm for all models (actor, critic, tstepencoder, and trajencoder) is smaller—staying below 1. Would this be a cause for concern? Could this be related to the positional embedding?

<img width="352" alt="Image" src="https://github.com/user-attachments/assets/bbff1824-4110-46d8-ba56-7c6fccc0e121" />

**Comment by jakegrigsby** [comment]
@Leiay I put my current progress on fixing this over on the the action_dists branch #78 for the time being. Running on my end again to triple check the -.15 number. Still getting to the bottom of the gap to .02.

**Comment by jakegrigsby** [comment]
> Running on my end again to triple check the -.15 number. This version with the command you posted above should be fine. 

Here's 3 runs while sweeping the exploration `steps_anneal` param over {[100k](https://wandb.ai/jakegrigsby/amagov3-public-popgym-regression/runs/8bpbdx9u?), [400k](https://wandb.ai/jakegrigsby/amagov3-public-popgym-regression/runs/72uf7k2d?), [1M](https://wandb.ai/jakegrigsby/amagov3-public-popgym-regression/runs/8juyy12h?)}. Looks fine. I'm doing 10M now just to check that parameter can ever matter. I'll check some old versions of the repo next.

**Comment by jakegrigsby** [comment]
This issue is caused by a brutal mistake where I apparently misread `rand` as `randn` [while refactoring random seeding in the exploration wrapper](https://github.com/UT-Austin-RPL/amago/commit/485e9cf930e3a89b88dc77f4ab0ba5d9f62e2d18) in mid november:

![Image](https://github.com/user-attachments/assets/219c21f0-6709-40af-84b9-adb64760014a)

So what appeared to be impacting one particularly hard env in one benchmark was actually changing every online exploration discrete action result. I guess this says something interesting about how the default hparams are probably overexploring in most easy tasks. Regardless I'll be pushing to the fix to #78 and merging after some final checks.

@Leiay thanks for brining this to my attention. I'm lucky this all happened between projects...

**Comment by jakegrigsby** [comment]
The fix has been merged to main and [here is a corrected reference run on MineSweeperMedium](https://wandb.ai/jakegrigsby/amagov3-public-popgym-regression/runs/ou8fxjvq).

![Image](https://github.com/user-attachments/assets/57e4ab8e-64f0-4fef-ac1f-8865c7a5e4b3)

**Comment by Leiay** [comment]
Hi @jakegrigsby 

I found that the performance is only reproducible if the "flash attention" option is on. See below for the two runs: the purple one is without flash attention installed, and the orange one is with flash attention installed.

<img width="402" alt="Image" src="https://github.com/user-attachments/assets/a39f0333-9570-4f1a-8b8c-b7bf04a604ca" />

That explains why even with the command I paste, I could not reproduce your performance up to -0.2. I was not using flash attention previously.

The difference between the runs with and without the flash attention is somehow significant, reflecting on all these training statistics.

<img width="1259" alt="Image" src="https://github.com/user-attachments/assets/a58f814a-3ab3-4113-bc57-8f1f8dc9cf35" />

I wonder why it is the case, since the flash attention is supposed to calculate the attention accurately. But in any case, maybe you could set the installation of flash attention by default.

**Comment by jakegrigsby** [comment]
@Leiay thanks I think i've fixed this. At least while I'm testing this I've removed the code that falls back to `VanillaAttention` when `flash_attn` isn't installed. I'll write more about this issue and clean things up later today.

**Comment by jakegrigsby** [comment]
- green: old commit from above issue
- orange: VanillaAttention (used when flash isn't installed)
- light blue: VanillaAttention with fixed pos embs
- brown: VanillaAttention with the fix
- grey: FlexAttention with the same problem fixed

<img width="427" alt="Image" src="https://github.com/user-attachments/assets/ff6e19dc-56cf-4d24-b87c-940526521bf5" />


also checked on RepeatPreviousEasy, RepeatFirstHard, T-Maze H=128, gymanx memory chain BSuite, cartpole, AutoencodeMedium

Still I've removed the choice to fall back to VanillaAttention when Flash isn't installed. I recommend using FlashAttention, and this bug only ever happened because I have flash installed and it's used in all of my tests.

**Comment by Leiay** [comment]
Thanks! Just out of curiosity, why using or not using flash attention could make such a huge difference?

**Comment by jakegrigsby** [comment]
The attention computation is not meaningfully different, but the key/value cache used during inference had a bug when each parallel actor (the batch dimension of the cache) is on a different timestep of its episode (and therefore writing to a different idx of the cache). This was covered during testing by coincidence because a lot of the easy memory tests I used to debug the VanillaAttention version reset based on time (every 200 timesteps for example), so every actor is on a consistent timestep. I didn't catch it since then because I always use flash attention in my day-to-day work.
